proposals/NNNN-typed-throws-concurrency-library.md

```diff
@@ -0,0 +1,381 @@
+
+```swift
+public struct ThrowingTaskGroup<ChildTaskResult: Sendable, Failure: Error> {
+  public mutating func next() async throws(Failure) -> ChildTaskResult?
+  public mutating func waitForAll() async throws(Failure)
+}
+
+struct ThrowingTaskGroup<ChildTaskResult: Sendable, Failure: Error> {
+  public struct Iterator: AsyncIteratorProtocol {
+    public mutating func next() async throws(Failure) -> Element? 
+  }
+}
+```
+
+### Task locals
+
+The `withValue` method can also adopt typed throws.
+
+```swift
+public final class TaskLocal<Value: Sendable> {
+  public func withValue<R, Failure: Error>(
+    _ valueDuringOperation: Value,
+    operation: () async throws(Failure) -> R,
+    file: String = #fileID,
+    line: UInt = #line
+  ) async throws(Failure) -> R 
+}
+```
+
+### MainActor
+
+The `MainActor` provides a static rethrowing `run` method. This method can also
+adopt typed throws:
+
+```swift
+public static func run<T: Sendable, Failure: Error>(
+  resultType: T.Type = T.self,
+  body: @MainActor @Sendable () throws(Failure) -> T
+) async throws(Failure) -> T
+```
+
+### `AsyncSequence` & `AsyncIteratorProtocol`
+
+`AsyncSequence` iterators can throw during iteration, as described by the
+`throws` on the `next()` operation on async iterators:
+
+```swift
+public protocol AsyncIteratorProtocol {
+  associatedtype Element
+  mutating func next() async throws -> Element?
+}
+```
+
+Introduce a new associated type `Failure` into this protocol to use as the
+thrown error type of `next()`, i.e.,
+
+```swift
+associatedtype Failure: Error = any Error
+mutating func next() async throws(Failure) -> Element?
+```
+
+Then introduce an associated type `Failure` into `AsyncSequence` that provides a
+more convenient name for this type, i.e.,
+
+```swift
+associatedtype Failure where AsyncIterator.Failure == Failure
+```
+
+With the new `Failure` associated type, async sequences can be composed without
+losing information about whether (and what kind) of errors they throw.
+
+With the new `Failure` type in place, we can adopt [primary asociated
+types](https://github.com/apple/swift-evolution/blob/main/proposals/0346-light-weight-same-type-syntax.md)
+for these protocols:
+
+```swift
+public protocol AsyncIteratorProtocol<Element, Failure> {
+  associatedtype Element
+  associatedtype Failure: Error = any Error
+  mutating func next() async throws(Failure) -> Element?
+}
+
+public protocol AsyncSequence<Element, Failure> {
+  associatedtype AsyncIterator: AsyncIteratorProtocol
+  associatedtype Element where AsyncIterator.Element == Element
+  associatedtype Failure where AsyncIterator.Failure == Failure
+  func makeAsyncIterator() -> AsyncIterator
+}
+```
+
+This allows the use of `AsyncSequence` with both opaque types (`some
+AsyncSequence<String, any Error>`) and existential types (`any
+AsyncSequence<Image, NetworkError>`). 
+
+### AsyncThrowingStream
+
+The non-throwing variant `AsyncStream` cannot adopt typed throws
```

Comment:
Can we not add a generic and give it a default?

Reply:
We can't add a new generic parameter to a type without breaking API or ABI AFAIK.